qlty-cloud/src/token/auth_flow.rs

```diff
@@ -0,0 +1,234 @@
+use super::DEFAULT_USER;
+use crate::token::{Token, SERVICE};
+use actix_web::{
+    dev::{Server, ServerHandle},
+    get,
+    http::header::LOCATION,
+    rt,
+    web::{Data, Query},
+    App, HttpResponse, HttpResponseBuilder, HttpServer,
+};
+use anyhow::{Ok, Result};
+use console::style;
+use serde::Deserialize;
+use std::{
+    net::TcpListener,
+    sync::{Arc, Mutex},
+    thread,
+    time::Duration,
+};
+use tracing::{debug, info, warn};
+use uuid::Uuid;
+
+const LOGIN_URL: &str = "https://qlty.sh/login";
+const LOGIN_URL_ENV: &str = "QLTY_AUTH_LOGIN_URL";
+
+#[derive(Deserialize, Debug)]
+struct AuthFlowQueryParams {
+    state: String,
+    code: String,
+    redirect_uri: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct AppState {
+    original_state: String,
+    server_handle: Arc<Mutex<Option<ServerHandle>>>,
+    credential_user: String,
+}
+
+impl Default for AppState {
+    fn default() -> Self {
+        AppState {
+            original_state: Uuid::new_v4().to_string(),
+            server_handle: Arc::new(Mutex::new(None)),
+            credential_user: DEFAULT_USER.to_string(),
+        }
+    }
+}
+
+#[get("/")]
+async fn auth_flow(
+    query: Query<AuthFlowQueryParams>,
+    state: Data<Arc<AppState>>,
+) -> HttpResponseBuilder {
+    if state.original_state != query.state {
+        warn!(
+            "Invalid state parameter in auth flow: {:?} did not match received state {:?}",
+            state.original_state, query.state
+        );
+        HttpResponse::BadRequest()
+    } else {
+        let token = Token::new(&state.credential_user);
+
+        match token.set(&query.code) {
+            Result::Ok(_) => {
+                debug!(
+                    "Auth token stored in credential storage: {}.{}",
+                    SERVICE, state.credential_user
+                );
+                stop_login_server(state);
+                HttpResponse::TemporaryRedirect()
+                    .insert_header((LOCATION, query.redirect_uri.clone()))
+                    .take()
+            }
+            Err(e) => {
+                warn!("Failed to store auth token in credential storage: {}", e);
+                HttpResponse::BadRequest()
+            }
+        }
+    }
+}
+
+fn login_server(state: Arc<AppState>) -> Result<(Server, u16)> {
```

Comment:
There's enough code related to the login server, and it is a distinct enough and important concept, that  I might extract it into its own file.

Reply:
Can you explain what you mean by this? This file is already pretty much entirely dedicated to the login server and handlers. We could move auth_via_browser out, which may be a source of confusion here.

Reply:
Resolving as I've moved the auth_via_browser to a separate file and what remains is all login-server-related.